File: src/系统/11．剧情系统/01．主线任务/02．剧情步骤/00．主线剧情/09．狩猎食人魔任务接取.lua
local ____lualib = require("lualib_bundle")
local __TS__Number = ____lualib.__TS__Number
local ____exports = {}
local ____03_FF0E_5267_60C5Boss_9884_7F6E_6865_63A5 = require("系统.11．剧情系统.01．主线任务.00．剧情系统核心工具.03．剧情Boss预置桥接")
local _____6CE8_518C_5267_60C5Boss_8303_56F4_9884_7F6E_89E6_53D1_5668 = ____03_FF0E_5267_60C5Boss_9884_7F6E_6865_63A5["注册剧情Boss范围预置触发器"]
local ____02_FF0E_5267_60C5NPC_521B_5EFA = require("系统.11．剧情系统.00．公共.02．剧情NPC创建")
local _____521B_5EFA_5267_60C5_573A_666F_5355_4F4D = ____02_FF0E_5267_60C5NPC_521B_5EFA["创建剧情场景单位"]
---
-- @noSelfInFile
local jass = require("jass.common")
local _____6C99_6F20_98DF_4EBA_9B54_5F85_6218_6682_505C_6765_6E90 = "剧情系统:沙漠食人魔待战"
local ____require_result_0 = require("系统.01．单位系统.08．单位配置表.02．Boss配置表")
local _____6309_540D_5B57_53CD_67E5Boss_5355_4F4DID = ____require_result_0["按名字反查Boss单位ID"]
local ____require_result_1 = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版")
local stringToFourCCSafe = ____require_result_1.stringToFourCCSafe
local ____require_result_2 = require("lib.扩展函数.BJ函数.01．触发与事件")
local TriggerRegisterUnitInRangeSimple = ____require_result_2.TriggerRegisterUnitInRangeSimple
local ____require_result_3 = require("系统.07．地形系统.03．区域传送")
local _____6CE8_518C_5267_60C5_914D_7F6E_4F20_9001 = ____require_result_3["注册剧情配置传送"]
local _____8BFB_53D6_5267_60C5_4F20_9001_914D_7F6E = ____require_result_3["读取剧情传送配置"]
local ____require_result_4 = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接")
local _____662F_73A9_5BB6_82F1_96C4_7EC4_5355_4F4D = ____require_result_4["是玩家英雄组单位"]
local _____83B7_53D6_73A9_5BB6_82F1_96C4_5355_4F4D_7EC4 = ____require_result_4["获取玩家英雄单位组"]
local ____require_result_5 = require("lib.扩展函数.BJ函数.08．单位BJ扩展")
local CreatePermanentCorpseLocBJ = ____require_result_5.CreatePermanentCorpseLocBJ
local ____require_result_6 = require("lib.扩展函数.Star扩展函数.GS扩展库.00．极坐标投影")
local GS_PolarProjectionBJ = ____require_result_6.GS_PolarProjectionBJ
local ____require_result_7 = require("lib.扩展函数.BJ函数.07．杂项")
local GetRandomDirectionDeg = ____require_result_7.GetRandomDirectionDeg
do
    local ____09_FF0E_72E9_730E_98DF_4EBA_9B54_4EFB_52A1_63A5_53D6 = require("系统.11．剧情系统.01．主线任务.02．剧情步骤.01．第一章.09．狩猎食人魔任务接取")
    ____exports["蛇人族接受食人魔任务剧情片段"] = ____09_FF0E_72E9_730E_98DF_4EBA_9B54_4EFB_52A1_63A5_53D6["蛇人族接受食人魔任务剧情片段"]
end
local GetUnitLoc = jass.GetUnitLoc
local Player = jass.Player
local RemoveLocation = jass.RemoveLocation
local PLAYER_NEUTRAL_PASSIVE = jass.PLAYER_NEUTRAL_PASSIVE
local bj_CORPSETYPE_BONE = require("jass.globals").bj_CORPSETYPE_BONE
local _____98DF_4EBA_9B54_4EFB_52A1_4F20_9001_914D_7F6EID = "jlc_desert_ogre_challenge"
local _____53D6_6D88_98DF_4EBA_9B54_4EFB_52A1_4F20_9001
local function _____6CE8_518C_98DF_4EBA_9B54_4EFB_52A1_4F20_9001()
    if _____53D6_6D88_98DF_4EBA_9B54_4EFB_52A1_4F20_9001 ~= nil then
        return
    end
    _____53D6_6D88_98DF_4EBA_9B54_4EFB_52A1_4F20_9001 = _____6CE8_518C_5267_60C5_914D_7F6E_4F20_9001(
        _____98DF_4EBA_9B54_4EFB_52A1_4F20_9001_914D_7F6EID,
        {
            ["读取玩家英雄组"] = _____83B7_53D6_73A9_5BB6_82F1_96C4_5355_4F4D_7EC4,
            ["允许进入单位"] = _____662F_73A9_5BB6_82F1_96C4_7EC4_5355_4F4D,
            ["完成"] = function()
                _____53D6_6D88_98DF_4EBA_9B54_4EFB_52A1_4F20_9001 = nil
            end
        }
    )
end
local function _____521B_5EFA_6C99_6F20_98DF_4EBA_9B54_5C38_9AA8_5708(bossUnit)
    if bossUnit == nil or bossUnit == 0 then
        return
    end
    local _____6B65_5175_5355_4F4DID = stringToFourCCSafe("hfoo")
    if not (_____6B65_5175_5355_4F4DID > 0) then
        return
    end
    do
        local i = 1
        while i <= 6 do
            local sourceLoc = GetUnitLoc(bossUnit)
            local corpseLoc = GS_PolarProjectionBJ(sourceLoc, 150, 60 * i)
            if corpseLoc ~= nil and corpseLoc ~= 0 then
                CreatePermanentCorpseLocBJ(
                    bj_CORPSETYPE_BONE,
                    _____6B65_5175_5355_4F4DID,
                    Player(PLAYER_NEUTRAL_PASSIVE),
                    corpseLoc,
                    GetRandomDirectionDeg()
                )
                RemoveLocation(corpseLoc)
            end
            i = i + 1
        end
    end
end
____exports["执行蛇人族接受食人魔任务"] = function(_____53C2_6570)
    local _____4F20_9001_914D_7F6E = _____8BFB_53D6_5267_60C5_4F20_9001_914D_7F6E(_____98DF_4EBA_9B54_4EFB_52A1_4F20_9001_914D_7F6EID)
    if _____4F20_9001_914D_7F6E ~= nil then
        local _____88C2_7F1D_5355_4F4D = _____521B_5EFA_5267_60C5_573A_666F_5355_4F4D({
            ["单位ID"] = "e08L",
            X = _____4F20_9001_914D_7F6E["入口中心X"],
            Y = _____4F20_9001_914D_7F6E["入口中心Y"],
            ["朝向"] = 0,
            ["玩家ID"] = PLAYER_NEUTRAL_PASSIVE,
            ["YD表"] = "剧情",
            ["YD键"] = "沙漠次元裂缝",
            ["YD字段"] = "unit"
        })
        if _____88C2_7F1D_5355_4F4D ~= nil and _____88C2_7F1D_5355_4F4D ~= 0 then
            _____6CE8_518C_98DF_4EBA_9B54_4EFB_52A1_4F20_9001()
        end
    end
    local bossRawId = _____6309_540D_5B57_53CD_67E5Boss_5355_4F4DID("沙漠食人魔")
    if bossRawId == nil then
        return
    end
    local bossUnit = _____521B_5EFA_5267_60C5_573A_666F_5355_4F4D({
        ["单位ID"] = bossRawId,
        X = 28354.9,
        Y = 13678.3,
        ["朝向"] = 270,
        ["玩家ID"] = PLAYER_NEUTRAL_PASSIVE,
        ["YD表"] = "Boss",
        ["YD键"] = "沙漠食人魔",
        ["YD字段"] = "unit",
        ["初始化无敌"] = true,
        ["初始化暂停来源"] = _____6C99_6F20_98DF_4EBA_9B54_5F85_6218_6682_505C_6765_6E90
    })
    if bossUnit == nil or bossUnit == 0 then
        return
    end
    _____6CE8_518C_5267_60C5Boss_8303_56F4_9884_7F6E_89E6_53D1_5668(
        bossUnit,
        __TS__Number(_____53C2_6570["注册范围"]) or 1000,
        "沙漠食人魔Boss启动",
        "jlc_desert_ogre_boss_start",
        "Boss.沙漠食人魔",
        10
    )
    _____521B_5EFA_6C99_6F20_98DF_4EBA_9B54_5C38_9AA8_5708(bossUnit)
end
____exports["狩猎食人魔任务接取剧情动作注册表"] = {["SRZ蛇人族_接受食人魔任务"] = ____exports["执行蛇人族接受食人魔任务"]}
return ____exports

File: TS/系统/11．剧情系统/01．主线任务/02．剧情步骤/00．主线剧情/09．狩猎食人魔任务接取.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const { 添加单位暂停 } = require("lib.扩展函数.Star扩展函数.Star扩展库.03．硬直暂停系统") as {
  添加单位暂停: (this: void, unit: any, source: string) => boolean;
};
const 沙漠食人魔待战暂停来源 = "剧情系统:沙漠食人魔待战";

const { YDUserDataSetSafe } = require("lib.扩展函数.YDWE函数.09．YDUserData安全版") as {
  YDUserDataSetSafe: (this: void, tableType: string, tableKey: any, attr: string, valueType: string, value: any) => void;
};
const { 按名字反查Boss单位ID } = require("系统.01．单位系统.08．单位配置表.02．Boss配置表") as {
  按名字反查Boss单位ID: (this: void, name: string) => string | undefined;
};
const { stringToFourCCSafe } = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版") as {
  stringToFourCCSafe: (this: void, s: string | undefined | null) => number;
};
const { TriggerRegisterUnitInRangeSimple } = require("lib.扩展函数.BJ函数.01．触发与事件") as {
  TriggerRegisterUnitInRangeSimple: (this: void, trig: any, range: number, whichUnit: any) => any;
};
const { 注册剧情配置传送, 读取剧情传送配置 } = require("系统.07．地形系统.03．区域传送") as {
  注册剧情配置传送: (this: void, 配置ID: string, 覆盖: {
    读取玩家英雄组: (this: void) => any;
    允许进入单位?: (this: void, unit: any) => boolean;
    完成?: (this: void, 触发单位?: any) => void;
  }) => (this: void) => void;
  读取剧情传送配置: (this: void, 配置ID: string) => {
    入口中心X: number;
    入口中心Y: number;
  } | undefined;
};
const { 是玩家英雄组单位, 获取玩家英雄单位组 } = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接") as {
  是玩家英雄组单位: (this: void, unit: any) => boolean;
  获取玩家英雄单位组: (this: void) => any;
};
const { CreatePermanentCorpseLocBJ } = require("lib.扩展函数.BJ函数.08．单位BJ扩展") as {
  CreatePermanentCorpseLocBJ: (this: void, style: number, unitid: number, whichPlayer: any, loc: any, facing: number) => any;
};
const { GS_PolarProjectionBJ } = require("lib.扩展函数.Star扩展函数.GS扩展库.00．极坐标投影") as {
  GS_PolarProjectionBJ: (this: void, source: any, dist: number, angle: number) => any;
};
const { GetRandomDirectionDeg } = require("lib.扩展函数.BJ函数.07．杂项") as {
  GetRandomDirectionDeg: (this: void) => number;
};

import type { 剧情动作参数表, 剧情动作处理器 } from "../../00．剧情系统核心工具/00．剧情动作类型";
import { 注册剧情Boss范围预置触发器 } from "../../00．剧情系统核心工具/03．剧情Boss预置桥接";
export { 蛇人族接受食人魔任务剧情片段 } from "../01．第一章/09．狩猎食人魔任务接取";

const CreateUnit = jass.CreateUnit as (this: void, owner: any, unitTypeId: number, x: number, y: number, facing: number) => any;
const GetUnitLoc = jass.GetUnitLoc as (this: void, whichUnit: any) => any;
const Player = jass.Player as (this: void, whichPlayer: number) => any;
const RemoveLocation = jass.RemoveLocation as (this: void, whichLocation: any) => void;
const SetUnitInvulnerable = jass.SetUnitInvulnerable as (this: void, whichUnit: any, flag: boolean) => void;

const PLAYER_NEUTRAL_PASSIVE = jass.PLAYER_NEUTRAL_PASSIVE as number;
const bj_CORPSETYPE_BONE = (require("jass.globals") as any).bj_CORPSETYPE_BONE as number;
const 食人魔任务传送配置ID = "jlc_desert_ogre_challenge";
let 取消食人魔任务传送: (() => void) | undefined;

function 注册食人魔任务传送(this: void): void {
  if (取消食人魔任务传送 != null) return;
  取消食人魔任务传送 = 注册剧情配置传送(食人魔任务传送配置ID, {
    读取玩家英雄组: 获取玩家英雄单位组,
    允许进入单位: 是玩家英雄组单位,
    完成: () => {
      取消食人魔任务传送 = undefined;
    },
  });
}

function 创建沙漠食人魔尸骨圈(this: void, bossUnit: any): void {
  if (bossUnit == null || bossUnit === 0) return;
  const 步兵单位ID = stringToFourCCSafe("hfoo");
  if (!(步兵单位ID > 0)) return;

  for (let i = 1; i <= 6; i++) {
    const sourceLoc = GetUnitLoc(bossUnit);
    const corpseLoc = GS_PolarProjectionBJ(sourceLoc, 150, 60 * i);
    if (corpseLoc != null && corpseLoc !== 0) {
      CreatePermanentCorpseLocBJ(bj_CORPSETYPE_BONE, 步兵单位ID, Player(PLAYER_NEUTRAL_PASSIVE), corpseLoc, GetRandomDirectionDeg());
      RemoveLocation(corpseLoc);
    }
  }
}

export function 执行蛇人族接受食人魔任务(this: void, 参数: 剧情动作参数表): void {
  const 传送配置 = 读取剧情传送配置(食人魔任务传送配置ID);
  const 次元裂缝单位ID = stringToFourCCSafe("e08L");
  if (传送配置 != null && 次元裂缝单位ID > 0) {
    const 裂缝单位 = CreateUnit(
      Player(PLAYER_NEUTRAL_PASSIVE),
      次元裂缝单位ID,
      传送配置.入口中心X,
      传送配置.入口中心Y,
      0,
    );
    if (裂缝单位 != null && 裂缝单位 !== 0) {
      YDUserDataSetSafe("string", "剧情", "沙漠次元裂缝", "unit", 裂缝单位);
      注册食人魔任务传送();
    }
  }

  const bossRawId = 按名字反查Boss单位ID("沙漠食人魔");
  const bossTypeId = stringToFourCCSafe(bossRawId);
  if (!(bossTypeId > 0)) return;
  const bossUnit = CreateUnit(Player(PLAYER_NEUTRAL_PASSIVE), bossTypeId, 28354.9, 13678.3, 270);
  if (bossUnit == null || bossUnit === 0) return;

  YDUserDataSetSafe("string", "Boss", "沙漠食人魔", "unit", bossUnit);
  SetUnitInvulnerable(bossUnit, true);
  添加单位暂停(bossUnit, 沙漠食人魔待战暂停来源);
  注册剧情Boss范围预置触发器(
    bossUnit,
    Number(参数.注册范围) || 1000,
    "沙漠食人魔Boss启动",
    "jlc_desert_ogre_boss_start",
    "Boss.沙漠食人魔",
    10,
  );
  创建沙漠食人魔尸骨圈(bossUnit);
}

export const 狩猎食人魔任务接取剧情动作注册表: Record<string, 剧情动作处理器> = {
  "SRZ蛇人族_接受食人魔任务": 执行蛇人族接受食人魔任务,
};
